sstables/sstable_set.hh

```diff
@@ -24,84 +24,207 @@
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable>&&);
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class sstable_set_snapshot;
+class version_reference;
+
+class sstable_list {
+    lw_shared_ptr<const std::unordered_set<versioned_sstable>> _versioned_sstables;
+    lw_shared_ptr<std::unordered_set<shared_sstable>> _unversioned_sstables;
+    uint64_t _size;
+    uint64_t _version;
+    friend class sstable_set;
+public:
+    sstable_list();
+    sstable_list(lw_shared_ptr<const std::unordered_set<versioned_sstable>> versioned_sstables, uint64_t size, uint64_t version);
+    sstable_list(const sstable_list& sstl);
+    sstable_list(sstable_list&& sstl) noexcept;
+    sstable_list& operator=(const sstable_list& sstl);
+    sstable_list& operator=(sstable_list&& sstl) noexcept;
+
+    class const_iterator : public std::iterator<std::forward_iterator_tag, const shared_sstable> {
+        std::unordered_set<versioned_sstable>::const_iterator _it;
+        std::unordered_set<versioned_sstable>::const_iterator _versioned_end;
+        std::unordered_set<shared_sstable>::const_iterator _unversioned_it;
+        uint64_t _version;
+    public:
+        const_iterator(std::unordered_set<versioned_sstable>::const_iterator it,
+                       std::unordered_set<versioned_sstable>::const_iterator versioned_end,
+                       std::unordered_set<shared_sstable>::const_iterator unversioned_it,
+                       uint64_t version);
+        const_iterator& operator++();
+        const_iterator operator++(int);
+        const shared_sstable& operator*() const;
+        bool operator==(const_iterator it) const;
+    };
+    using iterator = const_iterator;
+    const_iterator begin() const ;
+    const_iterator end() const;
+
+    size_t size() const;
+    void insert(shared_sstable sst);
+    void erase(shared_sstable sst);
+    bool contains(shared_sstable sst) const;
+    bool empty() const;
+};
+
+// should be owned by a lw_shared_ptr
+class sstable_set : public enable_lw_shared_from_this<sstable_set> {
     std::unique_ptr<sstable_set_impl> _impl;
     schema_ptr _schema;
     // used to support column_family::get_sstable(), which wants to return an sstable_list
     // that has a reference somewhere
-    lw_shared_ptr<sstable_list> _all;
-    std::unordered_map<utils::UUID, sstable_run> _all_runs;
+    lw_shared_ptr<std::unordered_set<versioned_sstable>> _all;
```

Comment:
Before the changes we could simply return _all in get_sstables() because for every version we had a separate copy of _all. We are avoiding copying so now we return the new struct sstable_list (which was also the name of the return type of get_sstables()) which iterates only over sstables belonging to the current version, instead of copying the set.